Add content to upgrade guide for v6 major bump (#208)
EasyPost/easypost-java: UPGRADE_GUIDE.md

Use the following guide to assist in the upgrade process of the `easypost-java` library between major versions.

- [Upgrading from 5.x to 6.0](#upgrading-from-5x-to-60)
- [Upgrading from 4.x to 5.0](#upgrading-from-4x-to-50)

## Upgrading from 5.x to 6.0

### 6.0 High Impact Changes

> - [Client Instance](#60-client-instance)

Reviewer: Question: Does `Client Object` make more sense here? I'm not married to either but a thought?

Author: I was following the C# upgrade guide to be consistent. What if we name the title to `Thread-safe`, I think this is more sense than `Client Instance/ Object`?

Author: Ok, let's leave it as `Client Instance`

- [Error Types](#60-error-types)

> - [Uniform empty response](#60-uniform-empty-response)

Reviewer: We know it as uniform empty responses, but let's say what it is from the users perspective. Please fix the links too

```suggestion
- [Empty Response Function Return Types](#60-uniform-empty-response)
```

### 6.0 Low Impact Changes

- [Services and Resources](#60-services-and-resources)

## 6.0 Client Instance

*Likelihood of Impact: **High***

> The library is now designed around the idea of a `EasyPostClient`. Users will initialize a `EasyPostClient` instance with their API key

Reviewer: ```suggestion
The library is now designed around the idea of an `EasyPostClient`. Users will initialize an `EasyPostClient` instance with their API key
```

and then use that instance to make API calls.

> This change allows for multiple clients to be instantiated with different API keys and for the library to be used in a
> multi-threaded environment.

Reviewer: ```suggestion
This change allows for multiple clients to be instantiated with different API keys which allows this library to be safely used in a
multi-threaded environment.
```

```java
// Old
EasyPost.apiKey = System.getenv("EASYPOST_API_KEY"); // global API key

// New
EasyPostClient client1 = new EasyPostClient("EASYPOST_API_KEY_1"); // per-client API key
EasyPostClient client2 = new EasyPostClient("EASYPOST_API_KEY_2"); // per-client API key
```

All functions are now accessed via the `EasyPostClient` instance. For example, to create a shipment:

```java
// Old
EasyPost.apiKey = System.getenv("EASYPOST_API_KEY");
Shipment shipment = Shipment.create(shipmentMap);

// New
EasyPostClient client = new EasyPostClient("EASYPOST_API_KEY_1");
Shipment shipment = client.shipment.create(shipmentMap);
```

## 6.0 Error Types

*Likelihood of Impact: **High***

Error handling has been overhauled and a number of specific exception types have been introduced.

All exceptions inherit from `EasyPost.Exception.EasyPostException` (which extends `System.Exception`). Users can catch this exception type to handle all errors thrown by the library.

> Subclasses of `EasyPostException` are grouped into two categories:

Reviewer: - Was all of this copied from the C# guide? 
- If so, are we sure we included every error type it had? 
- Did we double check that the syntax is correct for Java?

Author: Yes, it was copied from the C# guide but I have doubled check that every error type is included. I will remove the `EasyPost.Exception` namespace from the nested list since they are not necessarily needed.

  - `EasyPost.Exception.InvalidRequestError` - thrown when the API received an invalid request (422 status code)
  - `EasyPost.Exception.MethodNotAllowedError` - thrown when the API receives a request with an invalid HTTP method (
        405 status code)
  - `EasyPost.Exception.NotFoundError` - thrown when the API receives a request for a resource that does not exist (
        404 status code)
  - `EasyPost.Exception.PaymentError` - thrown when a payment error occurs (402 status code)
  - `EasyPost.Exception.ProxyError` - thrown when the library is unable to connect to the API via a proxy
  - `EasyPost.Exception.RateLimitError` - thrown when the API rate limit is exceeded (429 status code)
  - `EasyPost.Exception.RedirectError` - thrown when the http status is between 300 and 308.
  - `EasyPost.Exception.ServiceUnavailableError` - thrown when the API is unavailable (503 status code)
  - `EasyPost.Exception.TimeoutError` - thrown when the API request times out (408 status code)
  - `EasyPost.Exception.UnauthorizedError` - thrown when the API receives an unauthorized request (401 or 403 status
        code)

>   - `EasyPost.Exception.UnknownApiError` - thrown when an unknown API error occurs (unexpected 4xx status code)

Reviewer 2: Is the name space `Exception` or `Exceptions`? Inconsistent here, double-check throughout.

Author: Exception in Java

  - `EasyPost.Exceptions.FilteringError` - thrown when an error occurs during filtering (e.g. calculating lowest rate)
  - `EasyPost.Exceptions.InvalidObjectError` - thrown when an invalid object is being used
  - `EasyPost.Exceptions.InvalidParameterError` - thrown when an invalid parameter is being used
  - `EasyPost.Exceptions.MissingPropertyError` - thrown when a required property is missing (e.g. `Id` for most objects)
  - `EasyPost.Exceptions.SignatureVerificationError` - thrown when the signature for a webhook is invalid

Any exception thrown by the EasyPost library will be one of the above types.

Any `EasyPost.Exception.API` exception will have the following properties:

- `String Code` - the HTTP status code returned by the API call (e.g. 404)
- `String Message` - the error message returned by the API (e.g. "PARAMETER.INVALID")
- `List<Error> Errors` - a list of errors returned by the API (e.g. "Invalid parameter: to_address")

Users can better anticipate exception information through utilities in the `EasyPost.Exception.Constants` file.

## 6.0 Uniform empty response

*Likelihood of Impact: **High***

> Below functions return empty bodies or boolean have been changed to void.

Reviewer: ```suggestion
The following functions return an empty body from the API and have been changed to return void to reflect this.
```

- `fundWallet()` and `deletePaymentMethod()` in Billing class
- `createList()` in Tracker class
- `updateEmail()` in ReferralCustomer class

## 6.0 Services and Resources

*Likelihood of Impact: **Low***

> Static and instance-based methods have been divided into separate services and resources, methods that have API interactions are in services now. For example, the static method `Shipment.Create` is now accessible in the Shipment service via `client.Shipment.Create`. The instance method `myShipment.lowestRate` is still accessible only via a Shipment instance.

Reviewer: Shouldn't this be a lowercase `S`?

```suggestion
Static and instance-based methods have been divided into separate services and resources, methods that have API interactions are in services now. For example, the static method `Shipment.Create` is now accessible in the Shipment service via `client.shipment.Create`. The instance method `myShipment.lowestRate` is still accessible only via a Shipment instance.
```

Author: Yes, good catch

> Instances of an object cannot be initialized directly. Instead, use the service to create the instance via a `Create`, `Retrieve`, or `All` API call.

Reviewer: I'd argue that the likelyhood of this change is `high` because there are many function calls that are no longer instance-based. Please bump this up in the list and provide an example of something that used to be instance but is now static. eg: buying a shipment is no longer `shipment.buy()` but is now `client.shipment.buy("shp_...")`

Author: Resolved and splited the sentence to three bullet points

- `String Code` - the HTTP status code returned by the API call (e.g. 404)
- `String Message` - the error message returned by the API (e.g. "PARAMETER.INVALID")
- `List<Error> Errors` - a list of errors returned by the API (e.g. "Invalid parameter: to_address")

Users can better anticipate exception information through utilities in the `EasyPost.Exception.Constants` file.

## 6.0 Empty Response Function Return Types

*Likelihood of Impact: **High***

The following functions return an empty body from the API have been changed to return void.

- `fundWallet()` and `deletePaymentMethod()` in Billing class
- `createList()` in Tracker class
- `updateEmail()` in ReferralCustomer class

## 6.0 Services and Resources

*Likelihood of Impact: **High***

Static and instance-based methods have been divided into separate services and resources, methods that have API interactions are in services now.

For example:

- The static method `Shipment.buy()` is now accessible in the Shipment service via `client.shipment.buy("shp_...")`. 
- The instance method `myShipment.lowestRate()` is still accessible only via a Shipment instance because there is no API interaction.

> - The instance method of `refresh()` has been removed becasue you can no longer use object to make a HTTP request.

Reviewer: ```suggestion
- The instance method of `refresh()` has been removed because you can no longer use an object to make a HTTP request.
```

Instances of an object cannot be initialized directly. Instead, use the service to create the instance via a `create`, `retrieve`, or `all` API call.

## 6.0 Removal of Setters

*Likelihood of Impact: **Medium***

> Each object no longer has setters. The use case of the setter is extremely low, but if you need to use the setter for a specific reason, try to use `reflection` in Java.

Reviewer: ```suggestion
Objects no longer have setters. The use case of the setter is extremely low, but if you need to use the setter for a specific reason, try to use `reflection` in Java.
```

## 6.0 Rename Some Getters

*Likelihood of Impact: **Medium***

Below getters have been renamed:

- Pickup class: `getPickoutRates()` -> `getPickupRates()`
- PaymentMethod class: `getPrimaryPaymentMethodObject()` -> `getPrimaryPaymentMethod()`
- PaymentMethod class: `getSecondaryPaymentMethodObject()` -> `getSecondaryPaymentMethod()`

## 6.0 Type Changes

*Likelihood of Impact: **Low***

> Change the below type of objects:

Reviewer: ```suggestion
The following properties have changed types:
```
